tools/istio-iptables/pkg/cmd/run.go

```diff
@@ -236,17 +234,31 @@ func (iptConfigurator *IptablesConfigurator) handleInboundIpv6Rules(ipv6RangesEx
 		// ::6 is bind connect from inbound passthrough cluster
 		iptConfigurator.iptables.AppendRuleV6(constants.ISTIOOUTPUT, constants.NAT, "-o", "lo", "-s", "::6/128", "-j", constants.RETURN)
 
-		// Redirect app calls to back itself via Envoy when using the service VIP or endpoint
-		// address, e.g. appN => Envoy (client) => Envoy (server) => appN.
-		iptConfigurator.iptables.AppendRuleV6(constants.ISTIOOUTPUT, constants.NAT, "-o", "lo", "!", "-d", "::1/128", "-j", constants.ISTIOINREDIRECT)
-
 		for _, uid := range split(iptConfigurator.cfg.ProxyUID) {
+			// Redirect app calls back to itself via Envoy when using the service VIP
+			// e.g. appN => Envoy (client) => Envoy (server) => appN.
+			// nolint: lll
+			iptConfigurator.iptables.AppendRuleV6(constants.ISTIOOUTPUT, constants.NAT, "-o", "lo", "!", "-d", "::1/128", "-m", "owner", "--uid-owner", uid, "-j", constants.ISTIOINREDIRECT)
+
+			// Do not redirect app calls to back itself via Envoy when using the endpoint address
```

Comment:
interested to learn, why we need to not redirect when it's using endpoint ip? how does this help headless / mtls?

Reply:
For more context https://github.com/istio/istio/issues/18921#issuecomment-553385733